tp/tp.3
.\"/*+
.\"************************************************************************
.\"****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"*
.\"*	Name			: tp/doc/tp.3
.\"*
.\"*	Purpose
.\"*	    Contains the man page for the tp library.
.\"*
.\"*	Date			: March 19, 1999
.\"*
.\"*	SCCS data		: @(#)
.\"*	    Module Name		: tp.3
.\"*	    Version Number	: 1.1
.\"*	    Release Number	: 1
.\"*	    Last Updated	: 03/23/99
.\"*
.\"*	Written by		: Shannon Jaeger
.\"*
.\"*	Modification History:
.\"*
.\"****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"************************************************************************
.\"-*/
.
.de Ty
.ft 1
.fi
.RE
.PP
.ie n .TP 12n
.el .TP 16n
..
.
.\"			\" Td macro to display the type definition.
.de Td
.PP
.RS
.ft 3
.nf
.ie n .in 10n 53n
.ta 25n 60n
..
.
.\"			\" Ss macro to start a structure definition.
.de Ss
.PP
.RS
.nf
.ft 3
.ie n .ta 12n 26n 55n
.el .ta 16n 40n 70n
..
.\"			\" Es macro to end a structure definition.
.de Es
.ft 1
.fi
.RE
.PP
..
.\"
.TH tp 3 19/03/99 "CADC"
.SH NAME 
tp \- tape library routines.
.sp 1
.ta 5c 10c 10c 20c
tp_device_type	tp_init	tp_script_get
tp_type_get


.SH SYNOPSIS
.LP
.nf
.ft
#include "gen_types.h"
#include "gen_msg.h"
#include "tp.h"
.ft
.fi
.LP

.TP 7n
int 
tp_device_type( char * \fIdevice\fR, char * \fItype\fR);
.TP 7n
int 
tp_init( void );
.TP 7n 
int 
tp_script_get( char * \fItape_script_type\fR, TP_SCRIPTS ** \fIscript_names\fR); 

.SH DEFINITIONS
The following structure data types and global variables are defined in the
.B tp.h
header file.
.PP
.ti 2
.B STRUCTURES
.Ty
TP_SCRIPTS
This is a structure which contains the script names used to ingest, retrieve,
and name a particular tape script type.  
Ingesting is the process where the information found
on the tape is placed into an Archive database, retrieval is the extraction
of data off of the tape, and name detection involves determining the tapes
identifying name.  The members of the structure are:
.Ss
char	*tp_type;	/\(** The data type of the tape.	\(**/ 
char	*tp_ingest;	/\(** Ingestion script.		\(**/
char	*tp_retrieve;	/\(** Retrieval script.		\(**/
char	*tp_name;	/\(** Name extraction script.	\(**/
.Se
.PP
.ti 2
.B GLOBAL VARIABLES
.Ty
tp_msg
This is a character array which contains the success, warning, and/or
error messages that the 
.B tp
library reports.  See the 
.I gen_msg
UNIX man page for more information.


.SH DESCRIPTION
The \fBtp\fR library provides an interface to tape devices.

The library has its own configuration file,
.I libTp.config,
which contains all required configuration information.

The various library functions are briefly described below:
.Ty
tp_device_type(\|)
This function determines if a particular device is loaded and is a tape device.
The physical tape type is returned by this function in the \fItp_type\fR
parameter.
.Ty
tp_init()
Initializes the tape
.RB ( tp )
library.  This includes the reading of the 
.RB ( tp ) 
library's configuration file.
.Ty
tp_script_get()
Gets the script names for ingesting, retrieving, and determining the name 
determining for
a particular tape script type.  A tape script type uniquely determines the
scripts to use for ingestion, retrieval and name name determination.

.SH FILES
.SS libTp.config
The 
.B tp
library requires a configuration file which contains information
about tape scripts.  The file should be named
.I libTp.config
and located in a directory specified by a 
.B CONFIG_DIR
environment variable ( see the 
.B ENVIRONMENT VARIABLES
section ).  The ingest script and retrieve script values are mandatory; 
the name script value is optional.
The configuration file format is as follows:

.nf
.na
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 80n 88n 96n 104n 112n 120n 


#
# keyword       type	ingest_script	retrieve_script  name_script
# -------       ----    -------------   ---------------  -----------
#

type		CFHT1	tpIngestTar	tpRetrieveTar	 tpCfhtName
type		CFHT2	tpIngestTar	tpRetrieveTar
.fi


.SH ENVIRONMENT VARIABLES 
Environment variable  
.B LIBTP_CONFIG_DIR 
can be set to the path of a directory containing the configuration file.  If  
.B LIBTP_CONFIG_DIR 
is not set, the environment variable  
.B DEFAULT_CONFIG_DIR 
should be set to a path where the configuration file can be found. 

.SH DIAGNOSTICS
All of the  
.B tp
library functions return a 0 value upon
successful completion. Otherwise a non-zero number is returned.

.SH SEE ALSO
.BR gen_msg ( 3 ),
.BR mtio ( 7i )
on solaris, and
.BR mt ( 7 )
on hpux.
 
.SH BUGS 
On HP\-UX the mtio library is used to determine the tape device
type.  If the value of this library is not correct then the 
.B tp
library will not correctly determine the device type.  
DDS tape devices have been incorrectly identified as 8mm 
EXABYTE tape devices in the past.

.SH AUTHORS 
Shannon Jaeger
.br


National Research Council of Canada  
.br 
Herzberg Institute of Astrophysics 
.br 
Canadian Astronomy Data Centre 
		




CHANGES
==============================================
   DHS-lite sybase-less cfitio3.0.6 release
==============================================

gen++/genListOf.H                disable copy constructor
dhsStatus/staChannel.C           fix definition
dhsStatus/staChannel.H           fix missing public statement
dhsData/msg.C                    fix typo
dhsGen/dhsGenTmp.C               fix read_dir
dhsData/putFits.C                implement sybase-less option
dhsCommand/cmdCommandInfo.C      fix iterator declaration
dhsCommand/cmdDhs.H              fix friend declaration
conf/cxx_lib.mk                  fix clean target
conf/cc_bin.mk                   fix clean target
conf/cxx_bin.mk                  fix clean target
conf/cc_lib.mk                   fix clean target
conf/tcl.mk                      fix clean target
press/press.c                    add missing header
fh/fh.C                          precompiler switch for outdated cfitsio
dhsQlServer/qlsFrame.C           precompiler switch for outdated cfitsio
dhsQlServer/qlsDataset.C         precompiler switch for outdated cfitsio
sf/utils.c                       fix chunk size calculation, precompiler switch for outdated cfitsio
qldp/image.C                     fix Fits_IO.h header name under skycat-3.0.1
qldp/image.H                     fix Fits_IO.h header name under skycat-3.0.1
cdIngest/ingest.c                fix weird declaration?
sf/sds2fits.c                    leave libdd.config translation bug (would break GSA etc.)
                                 catch any attempt to move to an invalid hdu if maxHdr==0
                                 enforce max keys length to avoid HIERARCH convention
                                 precompiler switch for outdated cfitsio
conf/include.mk.in               several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
configure.in                     several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
dhsData/putSds.C                 critical fix for mem mapped fits
                                 precompiler switch for sybase-less dhs
dhsStatus/staMonitor.H           precompiler switch for sybase-less
dhsStatus/staMonitor.C           precompiler switch for sybase-less
dhsData/dtsDhs.C                 precompiler switch for sybase-less
dhsData/get.H                    precompiler switch for sybase-less
dhsData/get.C                    precompiler switch for sybase-less (needs improvement)
dhsStatus/Makefile               confitional flag for sybase-less
dhsData/Makefile                 confitional flag for sybase-less
dhsData/putFits.C                precompiler switch for sybase-less
dhsData/msg.C                    removing ending .
dhsData/dbm.C                    precompiler switch for sybase-less



tp/tp.c
/*+
************************************************************************
****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
*
* (c) 1995				(c) 1995.
* National Research Council		Conseil national de recherches
* Ottawa, Canada, K1A 0R6 		Ottawa, Canada, K1A 0R6
* All rights reserved			Tous droits reserves
* 					
* NRC disclaims any warranties,		Le CNRC denie toute garantie
* expressed, implied, or statu-		enoncee, implicite ou legale,
* tory, of any kind with respect	de quelque nature que se soit,
* to the software, including		concernant le logiciel, y com-
* without limitation any war-		pris sans restriction toute
* ranty of merchantability or		garantie de valeur marchande
* fitness for a particular pur-		ou de pertinence pour un usage
* pose.  NRC shall not be liable	particulier.  Le CNRC ne
* in any event for any damages,		pourra en aucun cas etre tenu
* whether direct or indirect,		responsable de tout dommage,
* special or general, consequen-	direct ou indirect, particul-
* tial or incidental, arising		ier ou general, accessoire ou
* from the use of the software.		fortuit, resultant de l'utili-
* 					sation du logiciel.
*
************************************************************************
*
*   Module Name:	tp/src/tp.c
*
*   Purpose:
*	Init the tplib, and get the struct of scripts.
*
*   Routines:
*	int	tp_device_type	: Check if the device is loaded and
*				: get the physical type of the tape.
*	int 	tp_init		: Init the tp lib.
*	int	tp_script_get	: Get the scripts.
*static int 	tp_type_compare_sort : Sort by type for eptr_search.
*
*   Date		: Feb 11, 1998
*
*   SCCS data		: @(#)
*	Module Name	: tp.c
*	Version Number	: 1.3
*	Release Number	: 1
*	Last Updated	: 04/12/99
*
*   Programmer		: George Zhang
*
*   Modification History:
*	99/03/23 SDJ	: Added hp support.
*	99/04/12 SDJ	: Removed extraneous debugging code.
*
****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
************************************************************************
-*/

#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <sybfront.h>
#include <sybdb.h>
#include <fcntl.h>

#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/mtio.h>

#include "gen_types.h"
#include "gen_str.h"
#include "gen_msg.h"
#include "gen_eptr.h"

#include "tp.h"
#include "local_tp.h"

static boolean  tp_init_flag;
EPTR		*tp_scripts_list;


/*+
************************************************************************
*
*   Function:	tp_device_type	(solaris version)
*
*   Purpose:
*	Get the physical type of the device.
*
*   Values Returned:
*	int	TP_E_DEVICE	: Error in device.
*	int	TP_SUCCESS	: All went well.
*
************************************************************************
-*/

/* XXX allan: replaced solaris/hpux #ifdefs with HAVE_MTDRIVETYPE check, updated for linux */
#ifdef HAVE_MTDRIVETYPE

/* Solaris version */

int	tp_device_type	
( 
   char		*device,	/* (in)  The device to check.	*/
   char		*tp_type	/* (out) The type of the device.*/
)
{
    int		dev_len; 	/* Length of the device name.		*/
    int		fd;		/* File descriptor.			*/
    struct mtdrivetype
		mtdt;		/* Tape Drive config. info.		*/	
    struct mtdrivetype_request	
		mtreq;		/* Tape Drive config. info. and size	*/	
    int		status;		/* Function return status.		*/
    

    /*
     *  Check the device.
     */

    dev_len = strlen( device );
    if ( device[dev_len - 1] != 'n' ||
       ( strncmp( "/dev/", device, 5 ) != 0 ) )
    {
	tp_format_message( TP_E_DEVICE, device );
	return( TP_E_DEVICE );
    }
     
    mtreq.size = sizeof( struct mtdrivetype );
    mtreq.mtdtp = &mtdt;
    fd = 0;
    status = TP_SUCCESS;

    
    if ( ( fd = open( device, O_RDONLY ) ) == -1 ||
	ioctl( fd, MTIOCGETDRIVETYPE, (char *) &mtreq ) == -1 )
    {
        switch( errno )
	{
	    case EPERM:
		tp_format_message( TP_E_PERM, device );
		break;

	    case EIO:
		tp_format_message( TP_E_UNLOAD, device );
		break;

	    case EBUSY:
		tp_format_message( TP_E_BUSY, device );
		break;

	    default:
		tp_format_message( MSG_ERRNO, device );
		break;
	}
				
	status = TP_E_DEVICE;
    }

    if ( fd >= 0 )
    {
	(void) close( fd );
    }
    TP_CHECK( status );


    /*
     *  Parse the type.
     */

    if ( strpattern( mtreq.mtdtp->name, TP_TYPE_DDS_PATTERN ) )
    {
	(void) strcpy( tp_type, TP_TYPE_DDS );
    }
    else if ( strpattern( mtreq.mtdtp->name, TP_TYPE_DLT_PATTERN ) )
    {
	(void) strcpy( tp_type, TP_TYPE_DLT );
    }
    else if ( strpattern( mtreq.mtdtp->name, TP_TYPE_EXA_PATTERN ) )
    {
	(void) strcpy( tp_type, TP_TYPE_EXA );
    }
    else
    {
	(void) strcpy( tp_type, TP_TYPE_UNKNOWN );
    }

    return( TP_SUCCESS );
}

#else /* HAVE_MTDRIVETYPE */

/* allan: XXX replaced hpux with linux version: added defines below */
#ifndef MT_ISDLT
#define MT_ISDLT 0x38
#endif

#ifndef MT_IS8MM
#define MT_IS8MM 0x35
#endif


/*+
************************************************************************
*
*   Function:	tp_device_type	(hpux version)
*
*   Purpose:
*	Get the physical type of the device.
*
*   Values Returned:
*	int	TP_E_DEVICE	: Error in device.
*	int	TP_SUCCESS	: All went well.
*
************************************************************************
-*/

int	tp_device_type	
( 
   char		*device,	/* (in)  The device to check.		*/
   char		*tp_type	/* (out) The type of the device.	*/
)
{
    int		    dev_len; 	/* Length of the device name.		*/
    int		    fd;		/* File descriptor.			*/
    struct mtget    mtgt;	/* Tape Drive property. info.		*/
    int		    status;	/* Function return status.		*/


    /*
     *  Check the device.
     */

    dev_len = strlen( device );
    if ( device[dev_len - 1] != 'n' ||
       ( strncmp( "/dev/", device, 5 ) != 0 ) )
    {
	tp_format_message( TP_E_DEVICE, device );
	return( TP_E_DEVICE );
    }

    fd = 0;
    status = TP_SUCCESS;

    
    if ( ( fd = open( device, O_RDONLY ) ) == -1 || 
	ioctl( fd, MTIOCGET, &mtgt ) == -1 )
    {
        switch( errno )
	{
	    case EPERM:
		tp_format_message( TP_E_PERM, device );
		break;

	    case EIO:
		tp_format_message( TP_E_UNLOAD, device );
		break;

	    case EBUSY:
		tp_format_message( TP_E_BUSY, device );
		break;

	    default:
		tp_format_message( MSG_ERRNO, device );
		break;
	}
	
	status = TP_E_DEVICE;
    }

    if ( fd >= 0 )
    {
	(void) close( fd );
    }
    TP_CHECK( status );
    

    /*
     *  Parse the type.
     */

    if ( mtgt.mt_type == MT_ISDDS1 || mtgt.mt_type == MT_ISDDS2 )
    {
	(void) strcpy( tp_type, TP_TYPE_DDS );
    }
    else if ( mtgt.mt_type == MT_ISDLT )
    {
	(void) strcpy( tp_type, TP_TYPE_DLT );
    }
    else if ( mtgt.mt_type == MT_IS8MM )
    {
	(void) strcpy( tp_type, TP_TYPE_EXA );
    }
    else
    {
	(void) strcpy( tp_type, TP_TYPE_UNKNOWN );
    }

    
    return( TP_SUCCESS );
}
#endif /* HAVE_MTDRIVETYPE */

/*+
************************************************************************
*
*   Function:   tp_init()
*
*   Purpose:
*	Initialize the tp library.
*
*   Values Returned:
*       int     TP_SUCCESS      : Normal completion.
*       int     TP_E_CONFIG     : Error in configuration.
*
************************************************************************
-*/

int 		tp_init
(
    void
)
{
    TP_CHECK( tp_config_init() );


    tp_init_flag = TRUE;

    return( TP_SUCCESS );
}

/*+
************************************************************************
*
*   Function:	tp_script_get 
*
*   Purpose:
*	Return a pointer to a structure in the list whose type 
*	matches the input type.
*	
*   Values Returned:
*	int	TP_SUCCESS	: Normal completion.
*	int	TP_E_INIT	: Do not init the tp lib.
*	int	TP_E_TYPE	: Type not in the list.
*
************************************************************************
-*/

int		tp_script_get
(
    char	*type,		/* (in)  The tar operation. 		*/
    TP_SCRIPTS 	**scripts	/* (out) A struct of TP_SCRIPTS.	*/
)
{
    int		index;          /* Index into eptr list.                */
    TP_SCRIPTS	*tmp;


    if ( ! tp_init_flag )
    {
	tp_format_message( TP_E_INIT );
	return( TP_E_INIT );
    }

    TP_CHECK_NULL_CONFIG( tmp = gen_alloc( sizeof( TP_SCRIPTS ) ) );
    tmp->tp_type = strsav( type );

    if ( eptr_search( tp_scripts_list, tmp, &index, 
    	    tp_type_compare_sort ) )
    {
        *scripts =  eptritem( tp_scripts_list, index );
    }
    else 
    {
	tp_format_message( TP_E_TYPE, type );
        return( TP_E_TYPE );
    }

    return( TP_SUCCESS );
}

/*+
************************************************************************
*
*   Function:   tp_type_compare_sort
*
*   Purpose:
*       eptr_search compare routine for type config ingo.
*
*   Values Returned:
*       int     >1      : String is greater than format config info.
*       int     0       : String matches the format config info.
*       int     <1      : String is less than format config info.
*
************************************************************************
-*/

int		tp_type_compare_sort
(
    TP_SCRIPTS	*script1,
    TP_SCRIPTS  *script2
)
{
    return( strcmp( script1->tp_type, script2->tp_type ) );
}
